include/deal.II/distributed/fullydistributed_tria.h

```diff
@@ -0,0 +1,237 @@
+#define dealii_fullydistributed_tria_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/smartpointer.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/template_constraints.h>
+#include <deal.II/base/timer.h>
+
+#include <deal.II/distributed/p4est_wrappers.h>
+#include <deal.II/distributed/tria_base.h>
+
+#include <deal.II/grid/tria.h>
+
+#include <boost/range/iterator_range.hpp>
+
+#include <functional>
+#include <list>
+#include <set>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#ifdef DEAL_II_WITH_MPI
+#  include <mpi.h>
+#endif
+
+#ifdef DEAL_II_WITH_P4EST
+#  include <p4est.h>
+#  include <p4est_connectivity.h>
+#  include <p4est_ghost.h>
+#  include <p8est.h>
+#  include <p8est_connectivity.h>
+#  include <p8est_ghost.h>
+#endif
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  namespace DoFHandlerImplementation
+  {
+    namespace Policy
+    {
+      template <typename>
+      class ParallelDistributed;
+    } // namespace Policy
+  }   // namespace DoFHandlerImplementation
+} // namespace internal
+
+namespace FETools
+{
+  namespace internal
+  {
+    template <int, int, class>
+    class ExtrapolateImplementation;
+  }
+} // namespace FETools
+
+// forward declaration of the data type for periodic face pairs
+namespace GridTools
+{
+  template <typename CellIterator>
+  struct PeriodicFacePair;
+}
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    struct CellInfo
+    {
+      CellInfo() = default;
+
+      CellInfo(CellId::binary_type index,
+               unsigned int        subdomain_id,
+               unsigned int        level_subdomain_id)
+        : index(index)
+        , subdomain_id(subdomain_id)
+        , level_subdomain_id(level_subdomain_id){};
+
+      CellId::binary_type index;
+      unsigned int        subdomain_id;
+      unsigned int        level_subdomain_id;
+    };
+
+
+
+    template <int dim, int spacedim>
+    struct ConstructionData
+    {
+      // information describing the local part of the coarse grid
+      std::vector<CellData<dim>>      cells;
+      std::vector<Point<spacedim>>    vertices;
+      std::vector<types::boundary_id> boundary_ids;
+
+      // information
+      std::map<int, int> coarse_lid_to_gid;
+
+      // information describing how to constuct the levels
+      std::vector<std::vector<CellInfo>> parts;
+    };
+
+
+
+    template <int dim, int spacedim = dim>
+    class Triangulation
+      : public parallel::DistributedTriangulationBase<dim, spacedim>
+    {
+    public:
+      typedef typename dealii::Triangulation<dim, spacedim>::cell_iterator
+        cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::active_cell_iterator
+          active_cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::CellStatus CellStatus;
+
+      void
+      reinit(ConstructionData<dim, spacedim> &construction_data);
```

Comment:
make the argument const (here and in many other places)

Reply:
Fixed here. More to come: WIP ;)